[SPARK-18191][CORE] Port RDD API to use commit protocol

jiangxb1987 opened the pull request:
## What changes were proposed in this pull request?

This PR port RDD API to use commit protocol, the changes made here:
1. Add new internal helper class that saves an RDD using a Hadoop OutputFormat named `SparkNewHadoopWriter`, it's similar with `SparkHadoopWriter` but uses commit protocol. This class supports the newer `mapreduce` API, instead of the old `mapred` API which is supported by `SparkHadoopWriter`;
2. Rewrite `PairRDDFunctions.saveAsNewAPIHadoopDataset` function, so it uses commit protocol now.

## How was this patch tested?
Exsiting test cases.

jiangxb1987 committed:
port RDD API to use commit protocol.

jiangxb1987 committed:
update comment.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -162,11 +162,14 @@ class SparkHadoopWriter(jobConf: JobConf) extends Logging with Serializable {
+  def createJobTrackerID(time: Date): String = {
+    new SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(time)
+  }
+
```
We need to generate jobTrackerID seprately in `SparkNewHadoopWriter `

jiangxb1987 commented on core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala:
```diff
@@ -509,21 +509,6 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
       (2, ArrayBuffer(1))))
   }
 
-  test("saveNewAPIHadoopFile should call setConf if format is configurable") {
```
This logic is no longer needed.

SparkQA commented:
**[Test build #68137 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68137/consoleFull)** for PR 15769 at commit [`e017e1e`](https://github.com/apache/spark/commit/e017e1e501429f55ebdbbf98f8aaab8f53902a40).

jiangxb1987 commented on core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala:
```diff
@@ -725,8 +710,7 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
 /*
-  These classes are fakes for testing
-    "saveNewAPIHadoopFile should call setConf if format is configurable".
+  These classes are fakes for testing saveAsHadoopFile/saveNewAPIHadoopFile.
```
This comment have been out-of-date for a while.

SparkQA commented:
**[Test build #68137 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68137/consoleFull)** for PR 15769 at commit [`e017e1e`](https://github.com/apache/spark/commit/e017e1e501429f55ebdbbf98f8aaab8f53902a40).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68137/
Test FAILed.

jiangxb1987 reviewed (commented):
I didn't touch `SparkHadoopWriter` and `PairRDDFunctions.saveAsHadoopDataset`, because they use the older `mapred` API, which is not supported by the `FileCommitProtocol` framework now.

jiangxb1987 committed:
fix scala style failure.

SparkQA commented:
**[Test build #68139 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68139/consoleFull)** for PR 15769 at commit [`5e12850`](https://github.com/apache/spark/commit/5e1285051c6ecc3020ee52b1f134f8c148bf2d65).

SparkQA commented:
**[Test build #68139 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68139/consoleFull)** for PR 15769 at commit [`5e12850`](https://github.com/apache/spark/commit/5e1285051c6ecc3020ee52b1f134f8c148bf2d65).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68139/
Test PASSed.

rxin commented on core/src/main/scala/org/apache/spark/SparkNewHadoopWriter.scala:
```diff
@@ -0,0 +1,133 @@
+ * OutputCommitter, is serializable.
+ */
+private[spark]
+class SparkNewHadoopWriter(
```
move this into internal/io

jiangxb1987 committed:
move SparkNewHadoopWriter to internal/io

SparkQA commented:
**[Test build #68168 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68168/consoleFull)** for PR 15769 at commit [`4e72745`](https://github.com/apache/spark/commit/4e72745da6e691d5f006184e85998e7519fab50b).

rxin commented:
Doesn't need to be in this PR, but can you also implement a HadoopMapRedCommitProtocol that supports the older mapred package's commiter?

SparkQA commented:
**[Test build #68168 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68168/consoleFull)** for PR 15769 at commit [`4e72745`](https://github.com/apache/spark/commit/4e72745da6e691d5f006184e85998e7519fab50b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68168/
Test FAILed.

jiangxb1987 committed:
move logic in saveAsNewAPIHadoopDataset into SparkNewHadoopWriter

jiangxb1987 committed:
bugfix

SparkQA commented:
**[Test build #68213 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68213/consoleFull)** for PR 15769 at commit [`09d5ed9`](https://github.com/apache/spark/commit/09d5ed906b5a178632e98f079fed65f5c3891706).

SparkQA commented:
**[Test build #68213 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68213/consoleFull)** for PR 15769 at commit [`09d5ed9`](https://github.com/apache/spark/commit/09d5ed906b5a178632e98f079fed65f5c3891706).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68213/
Test FAILed.

jiangxb1987 commented:
Moved the logic from `saveAsNewAPIHadoopDataset` into `SparkNewHadoopWriter`, also added `SparkNewHadoopWriterUtils` to store functions used by Hadoop/NewHadoop writer.
In a followup PR, I'll implement a similar protocol to support the older mapred package's committer.

jiangxb1987 committed:
bugfix

SparkQA commented:
**[Test build #68214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68214/consoleFull)** for PR 15769 at commit [`e5a60ff`](https://github.com/apache/spark/commit/e5a60ff66461045eb2a1754f41fd1412af5d76b0).

jiangxb1987 commented on core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala:
```diff
@@ -1092,37 +1093,38 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
+    // Instantiate writer
+    val committer = FileCommitProtocol.instantiate(
+      className = classOf[HadoopMapReduceCommitProtocol].getName,
+      jobId = stageId.toString,
```
In fact I'm not sure what value should be assigned to `jobId` here, should it be `jobTrackerId` combined with `jobId`, or any value else? I failed to find some code to follow on this topic.

SparkQA commented:
**[Test build #68214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68214/consoleFull)** for PR 15769 at commit [`e5a60ff`](https://github.com/apache/spark/commit/e5a60ff66461045eb2a1754f41fd1412af5d76b0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68214/
Test FAILed.

jiangxb1987 committed:
bugfix

SparkQA commented:
**[Test build #68215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68215/consoleFull)** for PR 15769 at commit [`86f1951`](https://github.com/apache/spark/commit/86f195199e336bbbadd9809dfb3923a9a948c195).

rxin commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -23,11 +23,11 @@ import java.text.SimpleDateFormat
 import org.apache.hadoop.mapred._
 import org.apache.hadoop.mapreduce.TaskType
 
 import org.apache.spark.internal.Logging
```
i'd move this file into spark/internal/io as well to be closer to SparkNewHadoopWriter.scala

rxin replied:
we can do it in a separate pr when you do the mapred committer.

jiangxb1987 replied:
Yes I'll do it later.

rxin commented on core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala:
```diff
@@ -0,0 +1,248 @@
+object SparkNewHadoopWriter extends Logging {
+
+  /** A shared job description for all the write tasks. */
+  private class WriteJobDescription[K, V](
```
maybe just remove this, since you have only 3 items here. In SQL there were a lot of items.

rxin commented on core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala:
```diff
@@ -0,0 +1,248 @@
+  def write[K, V: ClassTag](
+      sparkContext: SparkContext,
+      rdd: RDD[(K, V)],
+      committer: HadoopMapReduceCommitProtocol,
```
i'd move the creation of the commit protocol here here. The reason I put it outside in SQL was because streaming and batch needed to specify different protocols, but that problem doesn't exist in core.

SparkQA commented:
**[Test build #68215 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68215/consoleFull)** for PR 15769 at commit [`86f1951`](https://github.com/apache/spark/commit/86f195199e336bbbadd9809dfb3923a9a948c195).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68215/
Test PASSed.

jiangxb1987 committed:
refactor to improve readability.

SparkQA commented:
**[Test build #68219 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68219/consoleFull)** for PR 15769 at commit [`cfcd823`](https://github.com/apache/spark/commit/cfcd823225b3c1b0464b4b571c78ecf969a88f25).

SparkQA commented:
**[Test build #68219 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68219/consoleFull)** for PR 15769 at commit [`cfcd823`](https://github.com/apache/spark/commit/cfcd823225b3c1b0464b4b571c78ecf969a88f25).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68219/
Test FAILed.

jiangxb1987 commented:
The failed test case is imported by #15725, which is not related to our changes in this PR.

rxin commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala:
```diff
@@ -0,0 +1,242 @@
+
+    // Try to write all RDD partitions as a Hadoop OutputFormat.
+    try {
+      sparkContext.runJob(rdd, (context: TaskContext, iter: Iterator[(K, V)]) => {
```
we need to collect the result coming from the commit protocol here, and pass it into commitJob, don't we?

rxin commented:
The rest looks good.

cc @ericl for another look too

rxin commented:
Also @mridulm who has looked at one of the prs for commit protocol too.

@mridulm can you help review this too?

jiangxb1987 commented:
retest this please.

SparkQA commented:
**[Test build #68258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68258/consoleFull)** for PR 15769 at commit [`cfcd823`](https://github.com/apache/spark/commit/cfcd823225b3c1b0464b4b571c78ecf969a88f25).

jiangxb1987 committed:
pass the result from commit protocol to commitJob.

SparkQA commented:
**[Test build #68265 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68265/consoleFull)** for PR 15769 at commit [`243b8ba`](https://github.com/apache/spark/commit/243b8ba106dd3a604a9053612d9d38482792a3db).

SparkQA commented:
**[Test build #68258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68258/consoleFull)** for PR 15769 at commit [`cfcd823`](https://github.com/apache/spark/commit/cfcd823225b3c1b0464b4b571c78ecf969a88f25).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68258/
Test FAILed.

SparkQA commented:
**[Test build #68265 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68265/consoleFull)** for PR 15769 at commit [`243b8ba`](https://github.com/apache/spark/commit/243b8ba106dd3a604a9053612d9d38482792a3db).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68265/
Test FAILed.

jiangxb1987 commented:
@rxin I've addressed your comment. Seems the test case is still failing... @ericl

rxin commented:
I've disabled the test.

SparkQA commented:
**[Test build #3417 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3417/consoleFull)** for PR 15769 at commit [`243b8ba`](https://github.com/apache/spark/commit/243b8ba106dd3a604a9053612d9d38482792a3db).

SparkQA commented:
**[Test build #3417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3417/consoleFull)** for PR 15769 at commit [`243b8ba`](https://github.com/apache/spark/commit/243b8ba106dd3a604a9053612d9d38482792a3db).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

mridulm commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala:
```diff
@@ -0,0 +1,243 @@
+      outputMetricsAndBytesWrittenCallback.foreach { case (om, callback) =>
+        om.setBytesWritten(callback())
+        om.setRecordsWritten(recordsWritten)
+      }
```
This looks like a behavior change - metric's are getting updated even when exceptions are thrown.
Do it after Utils.tryWithSafeFinallyAndFailureCallbacks completes, not in finally

mridulm commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala:
```diff
@@ -0,0 +1,243 @@
+
+    if (SparkHadoopWriterUtils.isOutputSpecValidationEnabled(sparkConf)) {
+      // FileOutputFormat ignores the filesystem parameter
+      val jobFormat = format.newInstance
```
If it is Configurable, we should invoke setConf on it - the earlier code was doing this.
The tests for that seem to have been removed as well.
Any reason why ?

jiangxb1987 replied:
The logic is to create a `OutputCommitter` from the given hadoop conf, which is now handled by `HadoopMapReduceCommitProtocol`, so this code and corresponding tests are no longer needed.

mridulm replied:
I think I added the comment to the wrong location (though it is relevant here, it is probably less serious ?).
HadoopMapReduceCommitProtocol.setupCommitter() should be doing a setConf if OutputFormat is Configurable.

This needs to be fixed to ensure custom OutputFormat's work.
I see that the PR has already been committed - can you please file a bug and fix it ?
The test will also need to be re-added.

jiangxb1987 replied:
Oh...I see the problem now. Will add that in a follow up, thank you for clarifying.

mridulm commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala:
```diff
@@ -0,0 +1,243 @@
+    }
+  }
+
+  val RECORDS_BETWEEN_BYTES_WRITTEN_METRIC_UPDATES = 256
```
private and move it to top of the object ?

mridulm reviewed (commented):
Overall looks good, added a few comments though.

jiangxb1987 committed:
refactor.

jiangxb1987 commented:
@mridulm I've added most of your comments. Thank you!

SparkQA commented:
**[Test build #68330 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68330/consoleFull)** for PR 15769 at commit [`9380f91`](https://github.com/apache/spark/commit/9380f91281587867d8a630199c01c4263bc2e197).

SparkQA commented:
**[Test build #68330 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68330/consoleFull)** for PR 15769 at commit [`9380f91`](https://github.com/apache/spark/commit/9380f91281587867d8a630199c01c4263bc2e197).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68330/
Test PASSed.

rxin commented:
Thanks - merging in master.

rezasafi commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala:
```diff
@@ -0,0 +1,249 @@
+      hadoopConf: Configuration): Unit = {
+    // Extract context and configuration from RDD.
+    val sparkContext = rdd.context
+    val stageId = rdd.id
```
Is this accurate? Seems weird that satgeId is set to be equal to rdd.id. What is the commit protocol here?

jiangxb1987 replied:
It follows the previous behavior, what's your concern here?

rezasafi replied:
We had a test failure using Spark 2.2 that seems to happen after this commit. The failure didn't occur before using Spark 2.0. We thought there could be a problem since the jobContext passed to commiter.setupJob() (line 84) has a different JobID comparing to the task context that is passed to to commiter.setupTask() (line 126)  in the former, it comes from rdd.id and in the later, its from stage.id. Just wanted to check what is the protocol here that requires such a difference? Thanks.

rezasafi replied:
I investigated the problem a little more and I filed a jira & fix, see SPARK-22162. Thank you in advance.
